File: blind75/Two Sum/Two Sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        ### naive solution (brute force) - Time complexity is O(n^2)
        # indices = list(range(len(nums)))
        # for index1 in indices:
        #     for index2 in indices:
        #         if index1 == index2:
        #             continue
        #         elif nums[index1] + nums[index2] == target:
        #             output = [index1, index2]
        #             print(output)
        #             return sorted(output)
        # return []
        
        ### better solution
        tracking = dict()
        for i in range(len(nums)):
            complement = target - nums[i]
            if complement in tracking:
                return [tracking[complement], i]
            else:
                tracking[nums[i]] = i
        return []

File: blind75/Two Sum/Two Sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> tracking = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            int diff = target - nums[i];
            if (tracking.containsKey(diff)) {
                int[] output = {tracking.get(diff), i};
                return output;
            } else {
                tracking.put(nums[i], i);
            }
        }

        return new int[0];
    }
}
